Userland/Libraries/LibCompress/Zlib.cpp

```diff
@@ -68,16 +68,22 @@ u32 ZlibDecompressor::checksum()
     return m_checksum;
 }
 
-ZlibCompressor::ZlibCompressor(OutputStream& stream, ZlibCompressionLevel compression_level)
-    : m_output_stream(stream)
+ErrorOr<NonnullOwnPtr<ZlibCompressor>> ZlibCompressor::construct(OutputStream& stream, ZlibCompressionLevel compression_level)
 {
     // Zlib only defines Deflate as a compression method.
     auto compression_method = ZlibCompressionMethod::Deflate;
 
-    write_header(compression_method, compression_level);
+    auto zlib_compressor = TRY(adopt_nonnull_own_or_enomem(new (nothrow) ZlibCompressor(stream, compression_level)));
+    zlib_compressor->write_header(compression_method, compression_level);
+
+    return zlib_compressor;
+}
 
+ZlibCompressor::ZlibCompressor(OutputStream& stream, ZlibCompressionLevel compression_level)
+    : m_output_stream(stream)
     // FIXME: Find a way to compress with Deflate's "Best" compression level.
-    m_compressor = make<DeflateCompressor>(stream, static_cast<DeflateCompressor::CompressionLevel>(compression_level));
+    , m_compressor(make<DeflateCompressor>(stream, static_cast<DeflateCompressor::CompressionLevel>(compression_level)))
```

Comment:
Allocating the `DeflateCompressor` is fallible too, isn't it? I think that should be created outside and then passed in as a parameter.

Reply:
It indeed is, and I already kind of took care of it in a commit that I have scheduled down the line (we will switch to a fallible constructor due to switching to the new `CircularBuffer` implementation, which also needs to be allocated). It will presumably take some time until it lands in `master`, but can we keep it as-is until we get to the followup PR that is mentioned in #16758, which will contain that commit? Until then, this diff in particular doesn't make it worse than it currently already is, as far as I can tell.

Reply:
Sure, that's fine. :^)

Reply:
Turns out I'm stupid, I mixed up `ZlibCompressor` with `DeflateCompressor` due to the comment talking about allocating `DeflateCompressor` being fallible. `ZlibCompressor` already got a new and shiny `construct` function earlier in this PR, so I should have moved the allocation right then and there, so there is indeed no commit scheduled as of now. I'll fix this up as soon as we are done with the heavy changes.